---
title: "Amplicon Analysis in R"
output: 
  html_document:
    toc: true
    toc_float: true
    include:
      after_body: _footer.html
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This site contains lessons for analyzing 16S rRNA amplicon sequencing data in R based on data collected at the University of Wisconsin - Madison through the Micro 551 Senior Capstone class. Introduction to R materials were modified from the Data Carpentry course materials (Copyright (c) Data Carpentry) on [Data Analysis and Visualization in R for Ecologists](https://datacarpentry.org/R-ecology-lesson/). These lessons provide a brief overview of the experimental design, how sequences were preprocessed using [mothur](https://mothur.org/), and how to analyze the resulting OTU (operational taxonomical unit) table of abundances for each taxa in R.

All Data Carpentry instructional material is freely available under the Creative Commons Attribution [license](https://datacarpentry.org/R-ecology-lesson/LICENSE.html), and therefore this material is freely available under the same license.

## Overviews & Lessons:

1. [Project Overview](00-project-overview.html)
2. [Overview of Mothur Preprocessing](01-preprocessing.html)
3. [Introduction to R](02-intro-to-r.html)
4. [Analyzing Amplicon Data](03-analyzing-amplicon-data.html)

## Datasets

You will need to download the data files for the Introduction to R lesson and OTU tables for the Amplicon Analysis lesson. Downloading the raw sequencing files to perform preprocessing on your own is optional, and can be accessed through your UW Box Account.

- [Data files for Introduction to R materials](https://figshare.com/articles/Portal_Project_Teaching_Database/1314459)
- [OTU Tables for Amplicon Analysis Lessons]()
- [Raw Sequencing Files for Preprocessing](https://uwmadison.box.com/s/51uqumk335wde9lsj3zduwyo2mxasoyg) - you can only access these files if you have a UW-Madison Box account

## Installation Instructions

**R** and **RStudio** are separate downloads and installations. R is the
underlying statistical computing environment, but using R alone is no
fun. RStudio is a graphical integrated development environment (IDE) that makes
using R much easier and more interactive. You need to install R before you
install RStudio.

### Windows

#### If you already have R and RStudio installed

* Open RStudio, and click on "Help" > "Check for updates". If a new version is
	available, quit RStudio, and download the latest version for RStudio.
* To check which version of R you are using, start RStudio and the first thing
  that appears in the console indicates the version of R you are
  running. Alternatively, you can type `sessionInfo()`, which will also display
  which version of R you are running. Go on
  the [CRAN website](https://cran.r-project.org/bin/windows/base/) and check
  whether a more recent version is available. If so, please download and install
  it. You can [check here](https://cran.r-project.org/bin/windows/base/rw-FAQ.html#How-do-I-UNinstall-R_003f) for
  more information on how to remove old versions from your system if you wish to do so.

#### If you don't have R and RStudio installed

* Download R from
  the [CRAN website](https://cran.r-project.org/bin/windows/base/release.htm).
* Run the `.exe` file that was just downloaded
* Go to the [RStudio download page](https://www.rstudio.com/products/rstudio/download/#download)
* Under *Installers* select **RStudio x.yy.zzz - Windows
  Vista/7/8/10** (where x, y, and z represent version numbers)
* Double click the file to install it
* Once it's installed, open RStudio to make sure it works and you don't get any
  error messages.

### macOS

#### If you already have R and RStudio installed

* Open RStudio, and click on "Help" > "Check for updates". If a new version is
	available, quit RStudio, and download the latest version for RStudio.
* To check the version of R you are using, start RStudio and the first thing
  that appears on the terminal indicates the version of R you are running. Alternatively, you can type `sessionInfo()`, which will also display which version of R you are running. Go on
  the [CRAN website](https://cran.r-project.org/bin/macosx/) and check
  whether a more recent version is available. If so, please download and install
  it.

#### If you don't have R and RStudio installed

* Download R from
  the [CRAN website](https://cran.r-project.org/bin/macosx/).
* Select the `.pkg` file for the latest R version
* Double click on the downloaded file to install R
* It is also a good idea to install [XQuartz](https://www.xquartz.org/) (needed
  by some packages)
* Go to the [RStudio download page](https://www.rstudio.com/products/rstudio/download/#download)
* Under *Installers* select **RStudio x.yy.zzz - Mac OS X 10.6+ (64-bit)**
  (where x, y, and z represent version numbers)
* Double click the file to install RStudio
* Once it's installed, open RStudio to make sure it works and you don't get any
  error messages.


### Linux

* Follow the instructions for your distribution
  from [CRAN](https://cloud.r-project.org/bin/linux), they provide information
  to get the most recent version of R for common distributions. For most
  distributions, you could use your package manager (e.g., for Debian/Ubuntu run
  `sudo apt-get install r-base`, and for Fedora `sudo yum install R`), but we
  don't recommend this approach as the versions provided by this are
  usually out of date. In any case, make sure you have at least R 3.3.1.
* Go to the
  [RStudio download page](https://www.rstudio.com/products/rstudio/download/#download)
* Under *Installers* select the version that matches your distribution, and
  install it with your preferred method (e.g., with Debian/Ubuntu `sudo dpkg -i
  rstudio-x.yy.zzz-amd64.deb` at the terminal).
* Once it's installed, open RStudio to make sure it works and you don't get any
  error messages.


### Installing Packages

**After installing R and RStudio, you need to install the following packages.**

* After starting RStudio, at the console type these commands in order (you can copy and paste the following lines):

```
install.packages(c("tidyverse", "remotes"))
remotes::install_github("MadsAlbertsen/ampvis2")
if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
BiocManager::install("phyloseq")
```

If you are asked to update packages, type `a` for all. If you are asked to install packages from the source compilation, type `Yes`. You will know if the package installed successfully if you type `library(tidyverse)` and so on for each package and you do not get error messages. Note, when the tidyverse package is loaded, it will show some messages describing the package - this is not an error. 

* You can also do this by going to Tools -> Install Packages and typing the names of the packages separated by a comma.

## Resources

These materials were created by referencing several other resources, and this course website is by no means a completely exhaustive resource or tutorial on 16S rRNA amplicon analysis.